### PR TITLE
Initial prototype of CLI tool "@rushstack/trace-import"

### PR DESCRIPTION
**Pull request by octogonz**
## Summary

This is an initial prototype of a new CLI tool that helps with inspecting `node_modules` folders to understand installation problems, and also helps with understanding `import` statements that resolve incorrectly.

Installation will be like this:

```bash
npm install -g @rushstack/trace-import
```

## Details

CLI help looks like this:

```
trace-import 0.0.0 - https://rushstack.io

usage: trace-import [-h] [-d] -p MODULE_PATH [-b FOLDER_PATH] [-r {cjs,es,ts}]

This tool analyzes import module paths, to determine the resolved target
folder. For example, if the "semver" NPM package is installed, "trace-import
--path semver/index" will print output equivalent to the Node.js require.
resolve() API. If "@types/semver" is installed, then "trace-import
--resolution-type ts --path semver/index" will print the .d.ts file path that
would be resolved by a TypeScript import statement.

Optional arguments:
  -h, --help            Show this help message and exit.
  -d, --debug           Show the full call stack if an error occurs while
                        executing the tool
  -p MODULE_PATH, --path MODULE_PATH
                        The import module path to be analyzed. For example,
                        "example" in expressions such as: require("example");
                        require.resolve("example"); import { Thing } from
                        "example";
  -b FOLDER_PATH, --base-folder FOLDER_PATH
                        The "--path" string will be resolved as if the import
                        statement appeared in a script located in this folder.
                         If omitted, the current working directory is used.
  -r {cjs,es,ts}, --resolution-type {cjs,es,ts}
                        The type of module resolution to perform: "cjs" for
                        CommonJS, "es" for ES modules, or "ts" for TypeScript
                        typings. The default value is "cjs".
```

## How it was tested

This is an initial prototype:
- The algorithm is not expected to be 100% accurate right now. 
- For example, **package.json** path maps and **tsconfig.json** mappings are not implemented yet.  
- The `--resolution-type=es` mode (reading the `"module"` field from **package.json**) is not implemented at all yet.

You can test the build by running `rushx start` in the **apps/trace-import** folder.

You can also test by publishing locally:
```bash
cd apps/trace-import
pnpm pack
npm install -g rushstack-trace-import-0.0.0.tgz
```

### Resolving a CommonJS default index
```
C:\Git\rushstack\apps\trace-import>trace-import --path semver      

trace-import 0.0.0 - https://rushstack.io

Base folder:             C:\Git\rushstack\apps\trace-import
Package name:            semver
Module path:             (not specified)

Resolving...

Package folder:          C:\Git\rushstack\common\temp\node_modules\.pnpm\semver@7.3.8\node_modules\semver
package.json:            semver (7.3.8)
Default entry point:     "main": "index.js"

Target path:             C:\Git\rushstack\common\temp\node_modules\.pnpm\semver@7.3.8\node_modules\semver\index.js
```

### Resolving a CommonJS arbitrary path
```
C:\Git\rushstack\apps\trace-import>trace-import --path typescript/bin/tsc

trace-import 0.0.0 - https://rushstack.io

Base folder:             C:\Git\rushstack\apps\trace-import
Package name:            typescript
Module path:             bin/tsc

Resolving...

Package folder:          C:\Git\rushstack\common\temp\node_modules\.pnpm\typescript@4.8.4\node_modules\typescript
package.json:            typescript (4.8.4)

Target path:             C:\Git\rushstack\common\temp\node_modules\.pnpm\typescript@4.8.4\node_modules\typescript\bin\tsc
```

### Resolving a TypeScript declaration
```
C:\Git\rushstack\apps\trace-import>trace-import --resolution-type ts --path semver 

trace-import 0.0.0 - https://rushstack.io

Base folder:             C:\Git\rushstack\apps\trace-import
Package name:            semver
Module path:             (not specified)

Resolving...

Package folder:          C:\Git\rushstack\common\temp\node_modules\.pnpm\semver@7.3.8\node_modules\semver
package.json:            semver (7.3.8)
@types folder:           C:\Git\rushstack\common\temp\node_modules\.pnpm\@types+semver@7.3.5\node_modules\@types\semver
@types package.json:     @types/semver (7.3.5)
@types default index:    "types": "index.d.ts"

Target path:             C:\Git\rushstack\common\temp\node_modules\.pnpm\@types+semver@7.3.5\node_modules\@types\semver\index.d.ts
```

### CODE REVIEW

File: apps/trace-import/src/TraceImportCommandLine.ts [hunk 1]
```diff
@@ -0,0 +1,87 @@
+// Copyright (c) Microsoft Corporation. All rights reserved. Licensed under the MIT license.
+// See LICENSE in the project root for license information.
+
+import colors from 'colors';
```
Comment (iclanton) [inline apps/trace-import/src/TraceImportCommandLine.ts:4]:
```suggestion
import colors from 'colors/safe';
```

File: apps/trace-import/src/TraceImportCommandLine.ts [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+  CommandLineParser,
+  CommandLineFlagParameter,
+  CommandLineStringParameter,
+  CommandLineChoiceParameter
+} from '@rushstack/ts-command-line';
+import { InternalError } from '@rushstack/node-core-library';
+
+import { ResolutionType, traceImport } from './traceImport';
+
+export class TraceImportCommandLine extends CommandLineParser {
```
Comment (iclanton) [inline apps/trace-import/src/TraceImportCommandLine.ts:15]:
```suggestion
export class TraceImportCommandLineParser extends CommandLineParser {
```

Following convention in the other projects.

File: apps/trace-import/README.md [hunk 3]
```diff
@@ -0,0 +1,129 @@
...
+                        require.resolve("example"); import { Thing } from
+                        "example";
+  -b FOLDER_PATH, --base-folder FOLDER_PATH
+                        The "--path" string will be resolved as if the import
+                        statement appeared in a script located in this folder.
+                         If omitted, the current working directory is used.
+  -r {cjs,es,ts}, --resolution-type {cjs,es,ts}
+                        The type of module resolution to perform: "cjs" for
+                        CommonJS, "es" for ES modules, or "ts" for TypeScript
+                        typings. The default value is "cjs".
```
Comment (iclanton) [inline apps/trace-import/README.md:55]: Put this in a code block.

File: apps/trace-import/src/traceImport.ts [hunk 4]
```diff
@@ -0,0 +1,316 @@
...
+// See LICENSE in the project root for license information.
+
+import {
+  FileSystem,
+  IPackageJson,
+  IParsedPackageName,
+  JsonFile,
+  PackageName
+} from '@rushstack/node-core-library';
+import colors from 'colors';
```
Comment (iclanton) [inline apps/trace-import/src/traceImport.ts:11]:
```suggestion
import colors from 'colors/safe';
```

File: apps/trace-import/src/traceImport.ts [hunk 5]
```diff
@@ -0,0 +1,316 @@
...
+      console.log();
+      logOutputField('Target path:', dtsTargetPath);
+      return true;
+    }
+  }
+
+  return false;
+}
+
+function traceImport2(options: IExecuteOptions, warnings: string[]): void {
```
Comment (iclanton) [inline apps/trace-import/src/traceImport.ts:137]: 2?

Comment (iclanton) [reply apps/trace-import/src/traceImport.ts:137]:
```suggestion
function traceImportInner(options: IExecuteOptions, warnings: string[]): void {
```

**Review by iclanton — APPROVED** [review]

File: apps/trace-import/src/TraceImportCommandLine.ts (hunk 1 above)
Comment (octogonz) [reply apps/trace-import/src/TraceImportCommandLine.ts:4]: Good catch.  I'll add `Terminal` hopefully soon

File: apps/trace-import/src/TraceImportCommandLine.ts (hunk 2 above)
Comment (octogonz) [reply apps/trace-import/src/TraceImportCommandLine.ts:15]: Fixed

File: apps/trace-import/src/traceImport.ts (hunk 5 above)
Comment (octogonz) [reply apps/trace-import/src/traceImport.ts:137]: Improved the function names

**Pull request merged**